python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -985,6 +987,32 @@ def fetch_row_counts(
                 A set of corresponding Dagster events for dbt models, with row counts attached,
                 yielded in the order they are emitted by dbt.
         """
+        return self._attach_metadata(_fetch_row_count_metadata)
+
+    def _attach_metadata(
+        self,
+        fn: Callable[[DbtCliInvocation, DbtDagsterEventType], Optional[Dict[str, Any]]],
+    ) -> "DbtEventIterator[DbtDagsterEventType]":
+        """Runs a threaded task to attach metadata to each event in the iterator.
+
+        Args:
+            fn (Callable[[DbtCliInvocation, DbtDagsterEventType], Optional[Dict[str, Any]]]):
+                A function which takes a DbtCliInvocation and a DbtDagsterEventType and returns
+                a dictionary of metadata to attach to the event.
+
+        Returns:
+             Iterator[Union[Output, AssetMaterialization, AssetObservation, AssetCheckResult]]:
+                A set of corresponding Dagster events for dbt models, with any metadata output
+                by the function attached, yielded in the order they are emitted by dbt.
+        """
+
+        def _map_fn(event: DbtDagsterEventType) -> DbtDagsterEventType:
+            result = fn(self._dbt_cli_invocation, event)
+            if result is None:
+                return event
+
+            return event.with_metadata({**event.metadata, **result})
```

Comment:
One thing here: if we simplify the `fn` type, we'd expect users to call `.with_metadata`. 

I think a better name for `.with_metadata`  is `.replace_metadata`.

And for most users, they would call something like `.add_metadata`, which would call `.replace_metadata`, but tack on the `**event.metadata` on the user's behalf.